make GetPasswd take a terminal file descriptor

oconnor663 opened the pull request:
There are times when we need user input from the terminal, even though
our stdin is piped from something else. (For example, reading an
encrypted file over stdin, and asking the user for the password.) Make
GetPasswd\* take an explicit file descriptor for the terminal, so that
the caller has the option of using /dev/tty.

oconnor663 committed:
make GetPasswd take a terminal file descriptor

There are times when we need user input from the terminal, even though
our stdin is piped from something else. (For example, reading an
encrypted file over stdin, and asking the user for the password.) Make
GetPasswd* take an explicit file descriptor for the terminal, so that
the caller has the option of using /dev/tty.

oconnor663 commented:
Tested this on BSD and it turned out the build was broken. Needed a `C.int` cast. Fixed with 1179294.

maxtaco commented on win.go:
```diff
@@ -6,7 +6,7 @@ import "syscall"
 import "unicode/utf16"
 
-func getch() byte {
+func getch(termDescriptor int) byte {
```
Did @zanderz sign off on this?

maxtaco commented:
LGTM if @zanderz is ok with it

oconnor663 committed:
getchar() -> read() on BSD

The call to getchar() was still implicitly reading from stdin, rather
than the descriptor we were passing in.

oconnor663 commented:
A further BSD fix.

oconnor663 commented on bsd.go:
```diff
@@ -7,23 +7,23 @@ package gopass
+        tcsetattr(termDescriptor, TCSANOW, &t_new);
 
-        ch = getchar();
+        read(termDescriptor, &ch, sizeof(ch));
```
This is the main change in the latest commit. Also the whole function returns `char` instead of `int`. (If I read a single byte into a `sizeof(int)` buffer, I'll run into problems on big-endian platforms, right?)

maxtaco replied:
i'm not a fan of this change.  It's customary for `getch` to return int, since it needs to return `-1` on EOF. Some platforms have signed chars, and some have unsigned chars, so you can't count on `char` to convey `<0`

What's the argument against leaving `ch = getchar()` as is?

oconnor663 replied:
The problem with `getchar` is that it's implicitly reading from stdin, instead of from the file handle we want to pass it. What I want to do is "read a single byte from a file descriptor in the Most Standard Way." Do you know what way that is?

oconnor663 replied:
There's also `fgetc` and a bunch of other variants, but I think those require me to wrap the descriptor in a `FILE*`, and I don't know if we want to do that.

maxtaco replied:
Agreed that this library is somewhat jenky and not properly checking for error of `getch`.  But even still, I think we shouldn't ignore the error code from `read`...

zanderz commented:
LGTM, pending Max's comment :thumbsup:
